### ISSUE TITLE
[Feature Request]: Implement the new API of the filter-panel

### ISSUE THREAD

**Issue by gogonzo** [insightsengineering/teal.slice#138]
### Feature description

Current API can't support changes we are about to make. Thus, our api can't rely on the nested lists:
- each filter has a label
- filter-state made from citril.metadata contains a dataname attribute which previously wasn't a part of the specs (it was enough to put specs under dataname slot).
- We need a way to specify the filter group.
- We need to distinguish between interactive-filter (filter column) and fixed-filter (condition).

So, instead of:
```r
filters = list(
    ADSL = list(                                                          # FiletredDataset (data.frame) lvl
        <column> = list(selected, keep_na, keep_inf),                       # column level
        ...,
        filterable = c(<columns>)                                                   # attribute
    ),
    MAE = list(                                                           # FilteredDataset  (MAE) level
        patients = list(                                                     # subjects (DF) level
            <column> = list(selected, keep_na, keep_inf)                         # column level
        ),
        <experiment> = list(                                                 # experiment (SE, ...) level
            select = list(                                                       # colData level
                <column> = list(selected, keep_na, keep_inf),                       # column level
                ...
            ),
            subset = list(                                                       # rowData level
                <column> = list(selected, keep_na, keep_inf),                       # column level
                ...
            )
         ),
         ...
       )

    ),

)
```

I propose to specify filters through specific api-functions (they will produce lists):

1. Wrappers:
- filter_settings (for all filters) with arguments: ..., global, show_filter_manager, show_apply_filters. Returns `filter_settings` s3 class
- filter_var (so called interactive filter) with arguments: label, dataname, colname, selected, choices. Returns `filter`, `filter_variable` s3 classes
- filter_fixed (non-interactive filter) with arguments: id, label, dataname, condition. Returns `filter`, `filter_fixed` s3 classes
- filter_group with arguments: label, ... Returns `filter`, `filter_group` s3 classes
- most likely we need extra wrappers for MAE:
    - mae_patient_var(dataname, colname, selected, choices). Returns `filter`, `filter_mae_patient` s3 classes
    - mae_experiment_select(dataname, experimentname, colname, selected, choices). Returns `filter`, `filter_mae_select` s3 classes
    - mae_experiment_subset(dataname, experimentname, colname, selected, choices). Returns `filter`, `filter_mae_subset` s3 classes
    

2. Additional features/attributes of the filters:

- single/multiple
- filterable choices - to limit choices of the single filter. For example SEX to be only M and F. By default `choices = selected`

```r

filters = filter_settings(

    filter_var(dataname = "ADSL", varname = "ARM", choices = c("ARM A", "ARM B", "ARM C", NA)), # list

    filter_var(dataname = "ADSL", varname = "AGE", label = "AGE 18-60", choices = c(18, 60)),

    filter_fixed(id = "IT"),

    filter_fixed(label = "Custom label", dataname = "ADSL", condition = A == 1 & B %in% c(1, 2)), 

    filter_group(
        filter_fixed("MALE"),
        filter_var(dataname = "ADSL", varname = "AGE", c(18, 50, Inf, NA)
    ),

    mae_parient_var(dataname = "MAE", varname = "geneder"),

    mae_gene_subset(dataname = "MAE", experiment_name = "e1", varname = "gene.id"),

    mae_sample_var(),

   filterable = ...

  ),

   target = list(
     A = c("ARM A", "AGE 18-60", "IT", "MALE AGE"),
     B = c("ARM A", "IT"
     ...,
    )
)
```

API should also include customizations, like:
- `filter_settings(global = TRUE, ...)` disables module-specific-settings
- `filter_settings(show_filter_manager = FALSE)` doesn't show filter-manager
- `filter_settings(show_apply_filters = FALSE)` doesn't show the "+" gear which means that filter-panel will always have the same filters.

------------------------

Above filter_settings should be validated in FilteredData on each level.
- dataname of each filter should match available datasets
- colname within dataset should match colnames from this dataset
- selected should be within choices
- selected and choices should be withing values of the colname
- labels shouldn't be duplicated within one FilterStates

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines.

### Security Policy

- [X] I agree to follow this project's Security Policy.

**Comment by gogonzo** [comment]
@donyunardi @lcd2yyz @pawelru 
Please review the API proposition for the filter-panel extension. Instead of nested-lists I propose to make filters using wrappers.

**Comment by donyunardi** [comment]
Acceptance Criteria:
* Improve the current way to specify filters by not using nested list.
* Ability to create filter variable, fixed filter, filter group.
* Filter can be assign to specific module

Task: 
* We can add `filter_settings` and `filter_var` functionality
* When ready, add fixed filter and filter group

**Comment by gogonzo** [comment]
@danielinteractive @BFalquet 
What do you think about the idea of changing the API of the filter panel. There is a proposal to not specify filters with the nested list, but to use specific functions to create a new filter. For MAE, this would be:
- mae_parient_var(dataname = "MAE", varname = "gender", selected = ...),
- mae_gene_var(dataname = "MAE", experiment_name = "e1", varname = "xxx", selected = ...),
- mae_sample_var(dataname = "MAE", experiment_name = "e1", varname = "yyy", selected = ...),

We need these changes as we are going to introduce fixed_filter and filter_group, and operating on the `list()` would be confusing in my opinion.

**Comment by danielinteractive** [comment]
Thanks @gogonzo for looping us in - I think that could be good. Important as discussed before would be to make this modular, in the sense that extensions packages such as `teal.modules.hermes` and `teal.modules.helios` can define their own filter functions and that those can be used together with existing filter functions from the "base" teal packages.

**Comment by gogonzo** [comment]
@danielinteractive There is even more robust plan for you. We can move MAE stuff to the teal.modules.helios permanently, so there will be no need to overwrite.

**Comment by gogonzo** [comment]
Closing issue as it's not specific enough - part of the issue addressed in #187 
The remaining API issues will be created separately
